### ISSUE TITLE
Excessive file manipulation in SequentialRecordCache used in o.e.r.sail.nativerdf.TripleStore slows down bulk statement removal leading to transaction timeout exception

### ISSUE THREAD

**Issue by cadmiumkitty** [eclipse/rdf4j#3806]
### Current Behavior

Executing `connection.clear(<context>)` (and `CLEAR SILENT GRAPH <graph name>`) outside the transaction (also inside the transaction, but I don't have the right profiler snapshot to show it) results in a significant amount of time spent in o.e.r.common.io.NioFile.size() and o.e.r.common.io.NioFile.read() reading the records back from the SequentialRecordCache, even though the size or content of the cache are not changing.

### Expected Behavior

Faster `connection.clear(<context>)` (and `CLEAR SILENT GRAPH <graph name>`). It should be possible to tell SequentialRecordCache that writing to the cache is complete, so the size of the file can be cached, and records pre-read in bulk (perhaps a configuration option for number of records read ahead)

### Steps To Reproduce

1. Run Rdf4j server 3.7.7

2. Create a remote repository with RemoteRepositoryManager using the configuratoin snippet below:

```
  rep:repositoryImpl [
      rep:repositoryType "openrdf:SailRepository" ;
      sr:sailImpl [
          sail:sailType "openrdf:NativeStore" ;
          sail:connectionTimeOut "600000" ;
          ns:tripleIndexes "spoc,opsc,ospc" ;
          ns:forceSync false ;
          ns:valueCacheSize "10000000" ;
          ns:valueIDCacheSize "10000000" ;
          ns:namespaceCacheSize "10000";
          ns:namespaceIDCacheSize "10000" 
        ]
    ].
```

3. Import about 5K statements (in my case very simple SKOS taxonomy with about 1K concepts).

4. Run the following code in the client:

```
RepositoryConnection connection = ...
connection.setIsolationLevel(IsolationLevels.SNAPSHOT);
connection.begin();
connection.clear(...);
... some additional logic here...
connection.commit();
```

5. Observe operation taking about 10-20 seconds (running on a pretty old XPS-15 with Intel(R) Core(TM) i7-4712HQ CPU @ 2.30GHz and HDD with 4G RAM allocated to Rdf4j server)

6. Observe the profiler results that show that a lot of time is spent in TripleStore.removeTriples() working with SequentialRecordCache first writing deleted triples, then reading them back one by one always checking the file size.

![image](https://user-images.githubusercontent.com/8994424/163817005-f7420f1d-ec0d-42ec-8342-8119e4abc773.png)

![image](https://user-images.githubusercontent.com/8994424/163817032-a6a2cd14-52eb-4a93-9106-e10ead477740.png)

### Version

3.7.7

### Are you interested in contributing a solution yourself?

Perhaps?

### Anything else?

Not sure if this is really a bug or a feature request, but I am leaning towards a bug - on fairly small data sets (~60K records) transactions are timing out with exceptions. Similar performance issue on 4.0.0-M3

**Comment by hmottestad** [comment]
Could you try an explicit transaction using different isolation levels (in particular NONE)?

**Comment by cadmiumkitty** [comment]
Apologies, that specific profiler snapshot was within an explicit transaction at the SNAPSHOT level (I amended the client code in the original report). I experimented with different levels in and outside the transaction but the result is about the same (outside the transaction clear() takes about as much time as begin(), clear(), commit() within the transaction). I'll get the profiler snapshot for the NONE level and post here.

**Comment by cadmiumkitty** [comment]
The code to reproduce and profiler output with IsolatoinLevels.NONE: https://github.com/cadmiumkitty/rdf4j-issue-3806

One (significant) missing piece from the original issue description is that the problem is only visible when running in Docker with mapped volumes (taking just under 20 seconds versus just under 1 second when reading and writing directly to SSD to clear context with 10K records).

**Comment by cadmiumkitty** [comment]
It is significantly faster - for 10K records (this is for one sample run, I can get the profiler snapshot too if needed, I updated the code to reproduce the issue too):
[main] INFO com.dalstonsemantics.rdf4j.NativeStoreAdd - Creating statements.
[main] INFO com.dalstonsemantics.rdf4j.NativeStoreAdd - Statements created. Duration: 999
[main] INFO com.dalstonsemantics.rdf4j.NativeStoreClear - Calling connection.clear().
[main] INFO com.dalstonsemantics.rdf4j.NativeStoreClear - Completed connection.clear(). Duration: 395

Having said that, I still see the issue running a similar container as part of the Kubernetes set-up that does use named Docker volumes (appreciate this may be due to a specific StorageClass used and other parameters, but for this initial reference, Azure Premium_LRS is used).

**Comment by cadmiumkitty** [comment]
Another point to add (and I see that the initial commit on https://github.com/eclipse/rdf4j/compare/GH-3806-native-store-clear-slow will probably help with that too) is that simple bulk deletion of records as part of the SPARQL query as opposed to clear() on the connection is slow too in the same set up (I appreciate this would need to be logged as a separate issue, and I would need to get the numbers, etc.)

**Comment by hmottestad** [comment]
I've got a draft proposal for a fix here: https://github.com/eclipse/rdf4j/pull/3813

If you want to try it with docker you can check out the branch is use our docker build script:

https://github.com/eclipse/rdf4j/blob/develop/docker/build.sh

**Comment by cadmiumkitty** [comment]
It is taking me longer than expected to build the branch, I'll post the results here once I have everything running.

**Comment by cadmiumkitty** [comment]
Update based on the canned code to reproduce the issue - for 10K records the time to clear the graph went down from ~20s to ~500ms (updated code to reproduce the issue is in the same location https://github.com/cadmiumkitty/rdf4j-issue-3806):
13:16:22.191 [main] INFO com.dalstonsemantics.rdf4j.NativeStoreAdd - Creating statements.
13:16:27.628 [main] INFO com.dalstonsemantics.rdf4j.NativeStoreAdd - Statements created. Duration: 5437
13:16:27.760 [main] INFO com.dalstonsemantics.rdf4j.NativeStoreClear - Calling connection.clear().
13:16:28.334 [main] INFO com.dalstonsemantics.rdf4j.NativeStoreClear - Completed connection.clear(). Duration: 574

**Comment by cadmiumkitty** [comment]
Also to confirm there is about 98% speed-up when testing with real-world taxonomies with ~60K statements.

**Comment by hmottestad** [comment]
Fix will be available in the next minor release of RDF4J (4.1.0).
